Nvenc encoder

sgothel opened the pull request:
cherry picked patches from my '259-add-nvenc-encoder' branch, see https://github.com/sgothel/HandBrake/commits/259-add-nvenc-encoder 
applicable on official master branch

I have not picked @compninja28 patches regarding windows components
due to heavy collisions:
  - 3281077a9dbf51e6135f356d9969637086781af4
  - 634d2f7f487ed7f233cc271546a6240385f3c189
Hopefully @compninja28 can help re-merging them.

+++

NVENC: Remaining fixes for a 1st working draft on GNU/Linux (GTK)

Basically working:

    h264_nvenc
    hevc_nvenc
    Possible optimization:
    Even in constant quality mode, a user ceiling bit_rate might be desirable,
    currently only the maximum is used like in VP8/9: w x h x fps

Remaining issues:

    Muxer issue with h264_nvenc + MKV, ffprobe shows:

[h264 @ 0x55ae4dedacc0] sps_id 32 out of range
[h264 @ 0x55ae4dedacc0] Invalid nal size 1677734572
[h264 @ 0x55ae4dedacc0] missing picture in access unit with size 40644

Notable: This does not happen with hevc_nvenc

Tested:

    GNU/Linux Debian 9
    GeForce GTX 950
    BD 1080p source, cropped alignment 8: Around 90fps h264 + hevc cq 28, profile main

MatthewLazarow committed:
added h264_nvenc and hevc_nvenc encoder statements to module defs.

MatthewLazarow committed:
Added the nvenc entries to the switch statement just like VP8/9

MatthewLazarow committed:
Added nvenc encoders to dependency list

MatthewLazarow committed:
first attempt adding nvenc to common.c

MatthewLazarow committed:
added codec value to common header, updated work vcodec switch to try and mirror qsv_h264 values

MatthewLazarow committed:
modified the nvidia codecs in common code

MatthewLazarow committed:
fixed nvidia encoder names

MatthewLazarow committed:
updated module.defs to remove hwaccel

MatthewLazarow committed:
fixed callbacks.c name issue and added nvenc enable to moduel.defs

MatthewLazarow committed:
added entries to muxavformat and work files

MatthewLazarow committed:
updated internal and work files

MatthewLazarow committed:
Updated contrib ffmpeg module.defs to match gist provided by sr55

MatthewLazarow committed:
updated libhb/common.c values to match those provided by sr55 gist.

MatthewLazarow committed:
updated common.h to reflect values provided by sr55 gist

MatthewLazarow committed:
updated libhb encavcodec.c to match values provided by sr55 gist

MatthewLazarow committed:
updated libhb muxavformat.c to match values provided by sr55 gist.

MatthewLazarow committed:
updated ibhb/work.c to match values provided by sr55 gist.

MatthewLazarow committed:
fixed issue with empty space in module.def

MatthewLazarow committed:
added missed masks for nvenc codecs

sgothel committed:
NVENC: Use latest FFMPEG 3.4 working with NVENC

sgothel committed:
NVENC: Fix HB_VCODEC_FFMPEG_MASK covered values to lie within mask

sgothel committed:
NVENC: Fix HB_VCODEC_FFMPEG_H264 usage (gtk version missed a rename)

sgothel committed:
NVENC: Remaining fixes for a 1st working draft on GNU/Linux (GTK)

Basically working:
- h264_nvenc
- hevc_nvenc
Possible optimization:
- Even in constant quality mode, a user ceiling bit_rate might be desirable,
  currently only the maximum is used like in VP8/9: w x h x fps

Remaining issues:
1) Muxer issue with h264_nvenc + MKV, ffprobe shows:

[h264 @ 0x55ae4dedacc0] sps_id 32 out of range
[h264 @ 0x55ae4dedacc0] Invalid nal size 1677734572
[h264 @ 0x55ae4dedacc0] missing picture in access unit with size 40644

   Notable: This does not happen with hevc_nvenc

Tested:
- GNU/Linux Debian 9
- GeForce GTX 950
- BD 1080p source, cropped: Around 70fps h264 + hevc

sgothel committed:
NVENC: Fix VBR CQ Settings to be more adaptive

Adding 'rc-lookahead 32, init_qp 1, spatial_aq 1, aq-strength 8' for h264 and hevc
in case of VBR CQ mode, see
  <https://devtalk.nvidia.com/default/topic/1013152/video-technologies/crf-for-nvenc/post/5179877/#5179877>
and NVENC_VideoEncoder_API_ProgGuide.pdf v08 Chapter 8 p27, as well as Table 1 p33.

sr55 commented:
We currently use libav and don't have the capacity to be changing that short term, so we'll need to look at integrating any required fixes into libav. (or porting patches until such time libav pick them up)

I can handle any Windows UI components, so don't need to worry about that.

sr55 commented on libhb/encavcodec.c:
```diff
@@ -247,9 +313,35 @@ int encavcodecInit( hb_work_object_t * w, hb_job_t * job )
+            av_dict_set( &av_opts, "cq", quality, 0 );
+
+            // further Advanced Quality Settings in Constant Quality Mode
+            av_dict_set( &av_opts, "init_qpP", "1", 0 );
```
The UI's have an extra arguments box to allow strings to be passed through to configure the encoder. so we probably want to wire that in.

sgothel replied:
yes, we can add those .. additionally. At the commented point these are for proper adaptive constant quality like x264

sr55 commented on contrib/ffmpeg/module.defs:
```diff
@@ -9,31 +9,52 @@ endif
 
+# To build with nvenc: 
+# See <https://dwijaybane.wordpress.com/2017/07/19/ffmpeg-with-nvidia-acceleration-on-ubuntu-16-04-nvenc-sdk/>
+# 1) Copy NVENC SDK's Video_Codec_SDK_8.0.14/Samples/common/inc/*.h to /usr/local/include/
```
See, this is where I got a bit confused. Those headers I downloaded were NOT GPL compatible.   FFMpeg seems to inlclude the required headers in another folder to avoid needing the SDK for this reason.  libav currently does not.

sr55 replied:
Ah, this would be for libav right?

sr55 commented on contrib/ffmpeg/module.defs:
```diff
@@ -9,31 +9,52 @@ endif
+# See <https://dwijaybane.wordpress.com/2017/07/19/ffmpeg-with-nvidia-acceleration-on-ubuntu-16-04-nvenc-sdk/>
+# 1) Copy NVENC SDK's Video_Codec_SDK_8.0.14/Samples/common/inc/*.h to /usr/local/include/
+# 2) Build cudautils
+#    wget http://developer.download.nvidia.com/compute/redist/ffmpeg/1511-patch/cudautils.zip
```
What are these headers required for?

sgothel replied:
They are actually no more needed, since included in FFMPEG yup.
I copied these instruction earlier for myself .. should remove it.

sgothel committed:
NVENC: param typo: 'spatial_aq' is only used in hevc, h264 uses 'spatial-aq' :)

sgothel commented:
Depends on feature request 'issue' https://github.com/HandBrake/HandBrake/issues/974

jstebbins commented:
> See, this is where I got a bit confused. Those headers I downloaded were NOT GPL compatible. FFMpeg seems to inlclude the required headers in another folder to avoid needing the SDK for this reason. libav currently does not.

There has been a somewhat heated discussion over this lately on the ffmpeg mailing list.  The discussion was kicked off by the inclusion of external headers with the proposed AMD AMF patch. The thread below is the ongoing discussion after that patch was submitted.
https://ffmpeg.org/pipermail/ffmpeg-devel/2017-October/218834.html
https://ffmpeg.org/pipermail/ffmpeg-devel/2017-November/218932.html

The consensus seems to be that allowing this inclusion of external headers into the ffmpeg code base was a mistake.  But it's not clear yet what the outcome will be.

sgothel commented:
@sr55 @jstebbins w/o having read above convo, AFAIK current state of the art and law is that simple declarations in headers w/o any algorithms is fine to be included (or re-written).
However, it might get tricky with fancy algos/macros or poetry (names) .. 
As is .. it doesn't seem to be an issue and we are using FFMPEG's header inclusion of version 3.4 now.

sgothel committed:
NVENC: Removed old build instructions for Nvenc inclusion (SDK download, header inclusion etc). FFMPEG 3.4 includes them all.

sgothel committed:
NVENC: ffmpeg def: Remove slipped in --enable-libx264

commit 04668528d46ada39a72606457e8a868454e528c2 moving to ffmpeg 3.4
enabled it's libx264 codec. Removed.

sgothel commented:
@sr55 @jstebbins reading a bit .. looks like the compromise indeed is to include basic headers and to dynamically link against system libraries. Good.

sgothel commented:
I am w/ Timo here (include MIT licensed headers).. sometimes I believe discussions simply get out of hand.
Sure, if one wants to maintain and create a micro package for these headers, go ahead .. and one can use them later. But enforcing ppl to work hard for some ideal world grandslam solution is ridiculous.
Might be the lately SJW trend inducing too much noise? cough cough
Bottom line .. stay focused on the task and technical merits, while favoring clean solutions if feasible.

https://ffmpeg.org/pipermail/ffmpeg-devel/2017-October/218896.html

sgothel commented:
working on new branch on top of https://github.com/HandBrake/HandBrake/pull/981
and hence https://github.com/HandBrake/HandBrake/issues/974
